### ISSUE TITLE
Pass response contexts from RPC middleware to HTTP middleware

### ISSUE THREAD

**Issue by shunsukew** [paritytech/jsonrpsee#1509]
I am developing a rate-limiting RPC middleware. Currently, jsonrpsee framework utilizes `MethodResponse` Extensions in the RPC layer and `http::Response` Extensions in the HTTP layer, with these extensions operating independently of each other.

I would like to propagate context information from the RPC layer to the HTTP layer to achieve functionality such as returning a status code other than 200 (429 in my case) or, at the very least, adding an HTTP header to indicate rate limiting based on data in contexts.

What would be the best approach to transfer context (Extensions) information from a jsonrpsee RPC middleware to a hyper HTTP middleware?

**Comment by niklasad1** [comment]
I think we could copy the extensions from the MethodResponse to the HTTP response to fix that and then you would need some HTTP middleware to check whether some error occurred and set desired HTTP status code on the response.

This shouldn't be that hard to fix

**Comment by shunsukew** [comment]
That would be really helpful. Thank you so much.

**Comment by niklasad1** [comment]
> Addressing potential conflicts of the same Type data across calls within a batch is a subject for discussion.

Yeah, that one is tricky because the `Extensions` works based on the `rust type ids` and it's not possible to merge two types in a clean way I think i.e, just overwrite them with the most recent one...

I had in mind that the user could inject a Vec or another collection that each call RPC call/request can push to it but we can't do that internally in jsonrpsee without some hacks like merging the rust collection types ourselves such as `Vec<T>` and so on...

**Comment by shunsukew** [comment]
I think overwriting is good practical options under the current situation.
One idea I have is jsonrpsee only insert `RpcExtensions` type to Http middleware extensions regardless whether it is single or batch, where `RpcExtensions` is map of call id and `http::Extensions`.

**Comment by niklasad1** [comment]
>One idea I have is jsonrpsee introduces new definition RpcExtensions and only insert this type to Http middleware extensions regardless whether it is single or batch, where RpcExtensions is a map of call id and AnyMap.

> If users want to use AnyMap from jsonrpsee in hyper http layer, they can iterate over RpcExtensions.

Yes, that would work but I propose we go ahead with this simple approach first i.e, just overwrite but worth look into for a nicer API.

**Comment by shunsukew** [comment]
I think that's good to go. My local implmentation now depends on your [PR](https://github.com/paritytech/jsonrpsee/pull/1514) branch and working good.
